File: apps/fsharp/design-pattern/src/DesignPattern/Observer.fs
namespace DesignPattern.Observer

/// Observer パターン
/// F# ではコールバック関数のリストや Event 型で表現する。
module Observer =

    /// オブザーバーはメッセージを受け取る関数
    type Observer<'T> = 'T -> unit

    /// Subject はオブザーバーのリストを保持し、通知を行う
    type Subject<'T>() =
        let mutable observers: Observer<'T> list = []

        member _.AddObserver(observer: Observer<'T>) = observers <- observer :: observers

        member _.RemoveObserver(observer: Observer<'T>) =
            observers <- observers |> List.filter (fun o -> not (obj.ReferenceEquals(o, observer)))

        member _.NotifyObservers(value: 'T) =
            observers |> List.iter (fun observer -> observer value)

        member _.ObserverCount = observers.Length

    /// 給与を管理する従業員
    type Employee = { Name: string; mutable Salary: float }

    /// 従業員の給与変更を監視する Subject
    type EmployeeSubject() =
        let subject = Subject<string>()

        member _.AddObserver(observer) = subject.AddObserver(observer)
        member _.RemoveObserver(observer) = subject.RemoveObserver(observer)
        member _.ObserverCount = subject.ObserverCount

        member _.UpdateSalary(employee: Employee, newSalary: float) =
            let oldSalary = employee.Salary
            employee.Salary <- newSalary

            let message =
                sprintf "%s の給与が %.0f から %.0f に変更されました" employee.Name oldSalary newSalary

            subject.NotifyObservers(message)

File: apps/go/design-pattern/observer/observer.go
// Package observer demonstrates the Observer pattern in Go.
// Observers are callback functions stored in a slice.
package observer

import "fmt"

// Observer is a callback function that receives the employee after a change.
type Observer func(e *Employee)

// Employee is the subject that notifies observers when salary changes.
type Employee struct {
	Name      string
	Title     string
	Salary    float64
	observers []Observer
}

// NewEmployee creates a new Employee.
func NewEmployee(name, title string, salary float64) *Employee {
	return &Employee{
		Name:   name,
		Title:  title,
		Salary: salary,
	}
}

// AddObserver registers an observer.
func (e *Employee) AddObserver(o Observer) {
	e.observers = append(e.observers, o)
}

// RemoveObserver removes the last added observer.
func (e *Employee) RemoveObserver() {
	if len(e.observers) > 0 {
		e.observers = e.observers[:len(e.observers)-1]
	}
}

// SetSalary changes the salary and notifies all observers.
func (e *Employee) SetSalary(newSalary float64) {
	e.Salary = newSalary
	e.notifyObservers()
}

func (e *Employee) notifyObservers() {
	for _, o := range e.observers {
		o(e)
	}
}

// ObserverCount returns the number of registered observers.
func (e *Employee) ObserverCount() int {
	return len(e.observers)
}

// String returns a human-readable representation.
func (e *Employee) String() string {
	return fmt.Sprintf("%s (%s): %.0f", e.Name, e.Title, e.Salary)
}
